### ISSUE TITLE
Knowing which program is running

### ISSUE THREAD

**Issue by hasseily** [AppleWin/AppleWin#854]
It would be very useful to know which program is running on the emulator.

1. When booting, after a short enough time AppleWin calculates the running program's signature
2. It checks with a database of known signatures to determine the name of the program
3. Swapping disks after boot will not trigger a redetermination of the program. Whatever booted at start is what it is.

It is clear that calculating a common signature that is consistent across multiple cracks and image types is a potentially very complicated thing to do. I have started an attempt at that, and here is what I do:

- Upon `MemReset()` I create a `IORead_C00x()` trigger counter.
- I let `IORead_C00x()` trigger once because that's part of the boot process
- Upon the second `IORead_C00x()` trigger I assume enough of the program's code and data has been loaded to get a unique RAM signature. I also assume that this trigger will always happen at exactly the right place in the loaded program code
- Then I calculate for `mem` a CRC32 of:

```
// $08->$1F	(23 pages)
// $60->$BF (95 pages)
// for a total of 118 pages
```
The plan is then to match this array to a dictionary of known arrays of games and other software. Since we have 118 UINT32 values we can use neighbor algorithms to pick the right program, even if there was a slight difference in the code (unless the crack itself expects a keypress).

Unfortunately by default AppleWin initializes memory using `MIP_FF_FF_00_00` which randomizes a ton of memory pages. If I use `MIP_ZERO` I've got a decent shot at hitting the signature target.
What's the downside of using `MIP_ZERO`? Why initializing the memory with a weird pattern?

**Comment by sicklittlemonkey** [comment]
Ah, I see while writing this that you've found the init code in Memory.cpp:
The memory that changes between boots is probably the random init we do to
emulate the actual power-on pseudo-random patterns seen in the Apple II
DRAM. We looked at various physical machines, and the pattern is 00's and
FF's. There is often other randomness too, but anyway, we do this because
some games use memory locations as the seeds for their pseudo-random number
generators. See issue #206 - there are even screenshots.

I've thought about this on and off over the years. Hashing the disk images
was probably a better option, but the flies in the ointment are not only
(1) different cracks or images, but (2) saved game state or high scores,
and now (3) HDD images like Total Replay! It's hard to do anything about
(1) except link them somehow in the hash library. whereas (2) might be
overcome by hashing each track, keeping a byte for each, and only requiring
some minimum number of tracks match. (3) Brings us back to memory, but it's
hard to think of a heuristic that always works unless perhaps you keep
track of which memory has been read from disk.

**Comment by hasseily** [comment]
I'm looking right now at different options. I've got a solid foundation for it, going to be testing.

**Comment by hasseily** [comment]
And yes, sorry about the later editing of my comment. I was researching it and figured out the memory zeroing.
But why then is this done:
https://github.com/AppleWin/AppleWin/blob/fb0871672c0f8c1152238009b79834bbee8e4d32/source/Memory.cpp#L1822

What's the point, since main will be refilled anyway right after?

**Comment by Michaelangel007** [comment]
> Why initializing the memory with a weird pattern?

I implemented the _Memory Initialization Pattern_ (MIP).  `MIP_FF_FF_00_00 ` is pseudo-random because that IS what happens on real hardware. (Observed on my //e.) The comments have an hint at how you can verify this for yourself on real hardware:

```
	// Memory is pseudo-initialized across various models of Apple ][ //e //c
	// We chose a random one for nostalgia's sake
	// To inspect:
	//   F2. Ctrl-F2. CALL-151, C050 C053 C057
	// OR
	//   F2, Ctrl-F2, F7, HGR
```

If you are getting an access violation then your code broke something.

**Comment by Michaelangel007** [comment]
Providing some context. It is extremely useful to know _which_ software is currently running.  Which version of DOS or ProDOS, what Game or Program, etc. i.e. Automatic Software Identification has numerous uses.

At a future date at some point AppleWin will have an built-in `MemoryFingerprint()` function which will be needed for its "Auto Cheat" functionality. Bots (see #848), auto-mapping programs, etc. could also make use of this.

e.g.
```
// Scan both main and aux. memory generating a hash of each page which provides a "fingerprint" of the current running program/game/software.
// A hash value of 0 means no hash was generated for that page
// i.e. Used for the I/O page $C0, or for Aux. Memory which doesn't exist on an Apple ][+.
hash[] GetMemoryFingerprint()
```

If we assume a 48K application this means there are 256^49152 unique programs, or 6.468x 1 0^118369 -- for all practical purposes this is a near infinite number.  To reduce this down to a usable range we can "checksum" every 256 bytes of RAM. This "array of hash values" provides a "fingerprint" that can be used to identify the program.

Checksumming RAM into X bytes allows for small variations (game kracks) to not effect identification. This can even ben multi-threaded to make the identification process extremely fast.

**Comment by hasseily** [comment]
To put some hard numbers on @Michaelangel007 's comment, I've been experimenting all day with some CRCs.
First I used `MIM_ZERO` to remove the randomness, although this is not at all optimal. In the future we should keep track of which pages have been written to by the time of the first keystroke request, and only CRC those.
Then I CRC'ed every single page between $08 and $1f, and $60 and $8F.
I noticed that it was total overkill.
Now I'm at this stage:
```
// To calculate the program signature
// We will CRC32 the following pages of mem:
// $08
// $09->$1F (22 pages)
// $60->$8F (48 pages)
// $90->$8F (48 pages)
// for a total of 119 pages split in 4 CRCs
static void calculateProgramSig()
{
	g_ProgramSig[0] = calculateMemPageSig(0, 0x08);
	UINT crc = 0;
	for (UINT8 i = 0x09; i <= 0x1f; i++)
	{
		crc = calculateMemPageSig(crc, i);
	}
	g_ProgramSig[1] = crc;
	crc = 0;
	for (UINT8 i = 0x60; i <= 0x8f; i++)
	{
		crc = calculateMemPageSig(crc, i);
	}
	g_ProgramSig[2] = crc;
	crc = 0;
	for (UINT8 i = 0x90; i <= 0xBF; i++)
	{
		crc = calculateMemPageSig(crc, i);
	}
	g_ProgramSig[3] = crc;
}

static UINT calculateMemPageSig(UINT crc, UINT8 pageNumber)
{
	return crc32(crc, memmain + pageNumber * 256, 256);
}
```

It's very interesting and quite problematic at the same time. I noticed that with only the CRC of page 0x08 you could reliably identify multiple cracks of Ultima 5, and uniquely identify Sorcellerie 1,2 and 3, U4, U5, etc...
However one of my U4 dsk is a crack by Overbyte that has a crack screen, and that one has a different boot CRC of course.

The real problem with using only the boot CRC is that any different format is bound to be different. So the U4 WOZ version has a totally different CRC of boot (and the rest as well).

I think at this stage that we could say that program identification works on WOZ images. This way we know they're clean without any cracks and we just need to use the boot CRC. It's super fast and simple.

**Comment by hasseily** [comment]
I've been testing the simple page 0x08 CRC technique on a dozen games, and it holds up well.
The only place where it's special is that both Wizardry and Sorcellerie have the same CRC which in my opinion is a feature.
Now I'm pretty sure that any disk that boots DOS 3.3 or ProDOS will probably share its 0x08 CRC with every other one, I just need to download some images to test.

**Comment by hasseily** [comment]
Well everything I said above is wrong with respect to 0x08. I was looking at the hashes as read from another program which inadvertently inversed them. So in fact the CRC that works correctly is that of pages $90->$8F.
Which now brings us to the absolute need to track which pages are written to from boot, so we can go back to randomly initializing memory. I'll do that next.

**Comment by hasseily** [comment]
Okay I now have a new technique that seems to be showing promise:

As I wrote earlier, I stop everything I do related to figuring out the running program on the second poll for keypress.
During this time, any time UpdatePaging() is called, if the page is between 08 and C0 I put its page number in a 4-item array, popping the oldest number out.
So when the second keypress poll happens I have the ids of the last 4 written pages in main memory.
I then CRC each individually, and I have 4 CRC.

I have checked a dozen games for the CRC behavior, and I found that only Ultima V will have 2 of the CRC change randomly across restarts. All the others I tried are static and unique. Games with different cracks, as long as they don't expect a keypress at launch due to an intrusive crack screen, exhibit the same CRC.

I'm thinking of creating a dictionary where I hand-pick one CRC from the 4 for each game, and say that this is the canonical ID for the game. Then Applewin upon runtime can calculate the 4 CRCs and check each against the database. The first match will be the correct game.
If we find that one CRC isn't enough, we could use a combo of 2 for the ID.
